src/librustc_mir/interpret/intern.rs

```diff
@@ -319,10 +327,23 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
-            match base_intern_mode {
-                InternMode::Static => {}
-                InternMode::Const | InternMode::ConstBase => {
-                    // If it's not a static, it *must* be immutable.
+            match intern_kind {
+                // Mutable statics may contain mutable allocations even behind relocations
+                InternKind::Static(hir::Mutability::Mut) => {}
+                // Once we get heap allocations we need to revisit whether immutable statics can
+                // refer to mutable (e.g. via interior mutability) allocations.
+                InternKind::Static(hir::Mutability::Not) => {
+                    alloc.mutability = Mutability::Not;
+                }
+                // Raw pointers in promoteds may only point to immutable things so we mark
+                // everything as immutable. Creating a promoted with interior mutability is UB, but
```

Comment:
*rustc* decides to create a promoted, so this can never be UB. UB is what the *user* has to prove to us. What is their proof obligation here?

Reply:
Well... it would be UB if the user used that raw pointer to write to the memory allocation.

Reply:
So why is it correct to do prpmotion then? The same code must also be UB without the promotion, or else promotion is unsound.

Reply:
It is UB without and with promotion. All we do here is acknowledge this by making the allocation immutable so llvm puts it into immutable static memory. If the user modifies it at runtime, that's UB

Reply:
That doesn't answer my question. *Why* is it UB without promotion? Where is the UB coming from?

Reply:
Also this sounds like we are back to square 1 then: this is exactly what we used to do with `const`, right? We just marked them as immutable unconditionally.

Except now we do it for promoteds, which is much worse. The checks for promoteds should *never* be weaker than the checks for consts. When they are, that means one could write promoteds that desugar to consts that one couldn't write! That's basically unsoundness, as far as I am concerned.

Reply:
> Why is it UB without promotion? Where is the UB coming from?

mutating through a raw pointer obtained via an immutable reference

Reply:
> When they are, that means one could write promoteds that desugar to consts that one couldn't write!

Well... only if you were using the promoted in an unsound way. There's no user visible difference except for the fact that the user likely gets a segfault if they do UB vs they just write to semantically immutable (but not phyiscally immutable) memory.

Reply:
The only reason this is an issue at all here is that promoteds containing other promoteds don't create explicit promoteds in the `mir::Body`, but again use the "no StorageDead" trick

Reply:
Otherwise each promoted could at maximum contain one allocation without any nesting, and this one would be immutable by definition. But since we have the nested thing, the allocations need to be mutable so its value can be written.

Essentially

```
pub const FOO: &'static *const i32 = &(&0 as _);
```

is

```rust
pub const FOO: &'static *const i32 = {
    let x = 0;
    let y = &x;
    let z = y as _;
    &z
};
```

where all of `x`, `y`, `z` are temporaries. In order for the miri-engine being able to write to `x`, its allocation needs to be mutable.
Same for `z`, but since `z` is actually the base promoted, we just mark it as immutable without any checks. But then we have to recurse into its relocations to get to `x` and mark that.

I could try to figure out how to promote even inside promoteds. I think that's easy to do, but/since we explicitly do the opposite on purpose to reduce the number of promoteds.

Reply:
> The only reason this is an issue at all here is that promoteds containing other promoteds don't create explicit promoteds in the mir::Body, but again use the "no StorageDead" trick

Ah, that probably explains why when I tried to add an ICE in case there are mutable allocations in a `const` initializer, that failed -- back then promotion inside consts still did the "no `StorageDead`" thing.

> Otherwise each promoted could at maximum contain one allocation without any nesting, and this one would be immutable by definition.

So... can `const` even contain nested allocations currently?

Reply:
> mutating through a raw pointer obtained via an immutable reference

Please add a comment with this justification to the code.

Reply:
> So... can const even contain nested allocations currently?

Nope

Reply:
> Please add a comment with this justification to the code.

done

Reply:
> Nope

So... this is unreachable code, then? Might be better to just `bug!`, instead of discussing what we'd do hypothetically if could ever happen?

Reply:
This comment is talking about promoteds, which can contain nested allocations.